zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -335,47 +338,52 @@ where
-        // To avoid hangs, any computation in the critical section should
-        // be kept to a minimum.
-        let reconnect = {
-            let mut guard = self.address_book.lock().unwrap();
+        // Correctness: To avoid hangs, computation in the critical section should be kept to a minimum.
+        let address_book = self.address_book.clone();
+        let next_peer = move || -> Option<MetaAddr> {
+            let mut guard = address_book.lock().unwrap();
 
             // Now we have the lock, get the current time
             let instant_now = std::time::Instant::now();
             let chrono_now = Utc::now();
 
             // It's okay to return without sleeping here, because we're returning
             // `None`. We only need to sleep before yielding an address.
-            let reconnect = guard.reconnection_peers(instant_now, chrono_now).next()?;
+            let next_peer = guard.reconnection_peers(instant_now, chrono_now).next()?;
+
+            // TODO: only mark the peer as AttemptPending when it is actually used (#1976)
+            //
+            // If the future is dropped before `next` returns, the peer will be marked as AttemptPending,
+            // even if its address is not actually used for a connection.
+            //
+            // We could send a reconnect change to the AddressBookUpdater when the peer is actually used,
+            // but channel order is not guaranteed, so we could accidentally re-use the same peer.
+            let next_peer = MetaAddr::new_reconnect(&next_peer.addr);
+            guard.update(next_peer)
+        };
 
-            let reconnect = MetaAddr::new_reconnect(&reconnect.addr);
-            guard.update(reconnect)?
+        // Correctness: Spawn address book accesses on a blocking threadd, to avoid deadlocks (see #1976).
+        let next_peer = match tokio::task::spawn_blocking(next_peer).await {
+            Ok(next_peer) => next_peer?,
+            Err(task_panic) => panic!("panic in next peer task: {:?}", task_panic),
         };
 
-        // SECURITY: rate-limit new outbound peer connections
+        // Security: rate-limit new outbound peer connections
         sleep_until(self.min_next_handshake).await;
         self.min_next_handshake = Instant::now() + constants::MIN_PEER_CONNECTION_INTERVAL;
 
-        Some(reconnect)
+        Some(next_peer)
     }
 
     /// Mark `addr` as a failed peer.
     pub fn report_failed(&mut self, addr: &MetaAddr) {
         let addr = MetaAddr::new_errored(&addr.addr, addr.services);
         // # Correctness
         //
-        // Briefly hold the address book threaded mutex, to update the state for
-        // a single address.
-        self.address_book.lock().unwrap().update(addr);
+        // Spawn address book accesses on a blocking thread,
+        // to avoid deadlocks (see #1976).
+        let address_book = self.address_book.clone();
+        tokio::task::spawn_blocking(move || address_book.lock().unwrap().update(addr));
```

Comment:
TODO: await the spawned task
```suggestion
        tokio::task::spawn_blocking(move || address_book.lock().unwrap().update(addr));
```